service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+	for _, dbRule := range rules {
+		rule := dbRule.GetRule()
+		if dbRule.DeleteTimestamp == nil && isUnfiltered(rule) {
+			foundUnfiltered = true
+		}
+		tbid := rule.GetTargetBuildId()
+		if dbRule.DeleteTimestamp == nil && isInVersionSet(tbid, g.GetVersionSets()) {
+			return serviceerror.NewFailedPrecondition(fmt.Sprintf("update breaks requirement, target build id %s is already a member of version set", tbid))
+		}
+	}
+	if !foundUnfiltered {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+// isUnfiltered returns true if the assignment rule has no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func isUnfiltered(ar *taskqueue.BuildIdAssignmentRule) bool {
+	percentageRamp := ar.GetPercentageRamp()
+	return ar.GetHintFilter() == "" &&
+		ar.GetWorkerRatioRamp() == nil &&
+		(percentageRamp == nil || (percentageRamp != nil && percentageRamp.RampPercentage == 100))
+}
+
+// isInVersionSet returns true if the build id is in any of the listed version sets
+func isInVersionSet(tbid string, sets []*persistencepb.CompatibleVersionSet) bool {
+	for _, set := range sets {
+		for _, bid := range set.BuildIds {
+			if bid.GetId() == tbid {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func InsertAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_InsertBuildIdAssignmentRule,
+	maxARs int) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		data = &persistencepb.VersioningData{AssignmentRules: make([]*persistencepb.AssignmentRule, 0)}
+	} else {
+		data = common.CloneProto(data)
+	}
+	if req.GetRuleIndex() < 0 {
+		return nil, serviceerror.NewInvalidArgument("rule index cannot be negative")
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	slices.Insert(data.GetAssignmentRules(), int(req.GetRuleIndex()), &persistenceAR)
+	return data, checkAssignmentConditions(data, maxARs, false)
+}
+
+func ReplaceAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceBuildIdAssignmentRule) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		return nil, serviceerror.NewInvalidArgument("data cannot be nil for a replace call")
+	} else {
+		data = common.CloneProto(data)
+	}
+	rules := data.GetAssignmentRules()
+	idx := int(req.GetRuleIndex())
+	if len(rules) > idx+1 {
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(rules)))
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	slices.Replace(data.AssignmentRules, idx, idx+1, &persistenceAR)
```

Comment:
please note that user only sees the non-deleted rules and works with those indices. we cannot use the user-give index as is without taking deleted rules into account.

Reply:
fixed